sycl/plugins/unified_runtime/ur/adapters/hip/program.cpp

```diff
@@ -26,7 +106,122 @@ ur_result_t ur_program_handle_t_::setBinary(const char *Source, size_t Length) {
   return UR_RESULT_SUCCESS;
 }
 
+ur_result_t ur_program_handle_t_::finalizeRelocatable() {
+#ifndef SYCL_ENABLE_KERNEL_FUSION
+  assert(false && "Relocation only available with fusion");
+  return UR_RESULT_ERROR_UNKNOWN;
+#else
+  assert(IsRelocatable && "Not a relocatable input");
+  amd_comgr_data_t ComgrData;
+  amd_comgr_data_set_t RelocatableData;
+  if (UR_CHECK_ERROR(amd_comgr_create_data_set(&RelocatableData))) {
+    std::strcpy(ErrorLog,
+                "Failed to create comgr data set for the relocatable");
+    return UR_RESULT_ERROR_PROGRAM_BUILD_FAILURE;
+  }
+  COMgrDataSetTCleanUp RelocatableDataCleanup{RelocatableData};
+
+  if (UR_CHECK_ERROR(
+          amd_comgr_create_data(AMD_COMGR_DATA_KIND_RELOCATABLE, &ComgrData))) {
+    std::strcpy(ErrorLog, "Failed to create comgr data");
+    return UR_RESULT_ERROR_PROGRAM_BUILD_FAILURE;
+  }
+  // RAII for auto clean-up
+  COMgrDataTCleanUp DataCleanup{ComgrData};
+
+  if (UR_CHECK_ERROR(
+          amd_comgr_set_data(ComgrData, BinarySizeInBytes, Binary))) {
+    std::strcpy(ErrorLog, "Failed to create comgr data set");
+    return UR_RESULT_ERROR_PROGRAM_BUILD_FAILURE;
+  }
+
+  if (UR_CHECK_ERROR(amd_comgr_set_data_name(ComgrData, "jit_obj.o"))) {
```

Comment:
Is there any danger of name conflicts if we always use the same name here?

Reply:
We only use one object per data set and it is destroyed at the end of the scope.